overlord/snapstate/backend/link.go

```diff
@@ -140,6 +141,12 @@ func (b Backend) LinkSnap(info *snap.Info, dev snap.Device, linkCtx LinkContext,
 	// if anything below here could return error, you need to
 	// somehow clean up whatever updateCurrentSymlinks did
 
+	if restart != nil {
+		if err := restart.Restart(); err != nil {
+			logger.Noticef("WARNING: cannot restart services: %v", err)
```

Comment:
Previously when the restart happened inside generateWrappers() we would have returned an error. Not that it was handled in a reasonable way, nonetheless the first service to attempt a restart for was snapd, which was reasonable as failure to request it would mean that trying other services is pointless. Now, we're starting with snapd.apparmor, perhaps we should restore the original order, try restart snapd, then snapd.apparmor and the rest?

Also on error, we're in a bit more trouble as we have updated the current symlink, so this needs to be reverted somehow (even though the service files themselves encode snapd revision IIRC. Fortunately snapd snap does not participate in boot, so at least we don't have to revert anything that SetNextBoot did().

Reply:
`undoLinkSnap` will call `LinkSnap` on the previous version if there is an error and if the snap is snapd.

Reply:
I will test failures in the spread test.

Reply:
Right, I do not return any error here...

Reply:
The restart is non-blocking. So we do not necessarily catch all errors here.

Reply:
> The restart is non-blocking. So we do not necessarily catch all errors here.

That's fine, it's similar to what we had before. TBH the check was really to see if systemd even attempted to start a transaction. IIRC this was at most able to verify that the unit is valid (syntax wise)

Reply:
> Now, we're starting with snapd.apparmor, perhaps we should restore the original order, try restart snapd, then snapd.apparmor and the rest?

snapd.apparmor was restarted before snapd. All services without  `X-Snapd-Snap: do-not-start` were started before.

Reply:
So to sum up for the re-review:
 * Catching error is not that useful here, it is likely not to fail here because it is non blocking.
 * snapd-failure was broken, and I have pushed a fix.
 * I have added in the test a broken snapd to make sure the symlink gets reverted by snapd-failure
 * I do not think I have changed the order of restart.

Reply:
I'm still confused why we should return an error here even if it's rare?  we will not get restarted so we continue in the current snapd but the symlink is wrong, that seems odd

Reply:
The second case, I meant: If `snapd.service` does not fallback to `snapd.failure.service`, *and `snapd` fails to execute* we will not catch it.

Reply:
I will try to make a spread that can test an error there and see what happens if we revert the symlink and return and error.

Reply:
I have observed that if `snapd.service` is complete garbage. Then the `daemon-reload` that happens earlier will kill the currently snapd, and this code will not be reached. I will try to see what happens if `snapd.apparmor.service` or other service is garbage and see if we can cause an error here.